_posts/2016-05-11-android-testing-with-kotlin.md

```diff
@@ -0,0 +1,422 @@
+the strings lives in
+[`Formatter.java`](https://github.com/GreenhouseCI/kotlin-tests/blob/master/app/src/main/java/com/greenhouseci/kotlin_tests/kotlintests/Formatter.java)
+and looks like
+
+```java
+public class Formatter {
+
+    public String stripMiddle(String s) {
+        return s.replaceAll("\\s+", " ");
+    }
+
+    public String stripLeft(String s) {
+        return s.replaceAll("^\\s+", "");
+    }
+
+    public String stripRight(String s) {
+        return s.replaceAll("\\s+$", "");
+    }
+
+    public String strip(String s) {
+        return stripLeft(stripRight(stripMiddle(s)));
+    }
+}
+```
+
+Bridge connecting user interactions with formatter resides conveniently in
+[`MainActivityFragment`](https://github.com/GreenhouseCI/kotlin-tests/blob/master/app/src/main/java/com/greenhouseci/kotlin_tests/kotlintests/MainActivityFragment.java#L38).
+
+All in all, our app stands in it's full glory as depicted below:
+<div style="text-align:center; padding: 20px 0;"><img src="/assets/kotlin-app-screenshot.png" alt="Screenshot"/></div>
+
+Evidently we want to test out that whenever item description `EditText` is filled with anything
+meaningful and "Add item" button is pressed, the newly added item is populated to the list.
+Furthermore, it is of utmost importance to ensure that all the items that are added to the list
+do not contain any superfluous whitespace characters. For that we need tests of course.
+
+## Unit tests
+
+Let us start off with unit tests for our `Formatter` class shown above. As already mentioned,
+we'll store our unit tests under `src/test/kotlin`. Right there we have a test class
+`FormatterTestKotlin` defined in
+[`FormatterTestKotlin.kt`](https://github.com/GreenhouseCI/kotlin-tests/blob/master/app/src/test/kotlin/com/greenhouseci/kotlin_tests/kotlintests/FormatterTestKotlin.kt):
+
+```ruby
+package com.greenhouseci.kotlin_tests.kotlintests
+
+import org.junit.Assert.assertEquals
+import org.junit.Test
+
+
+class FormatterTestKotlin {
+
+    private val formatter: Formatter = Formatter()
+
+    infix fun Any.equals(expected: Any) {
+        assertEquals(expected, this)
+    }
+
+    ...
+
+    @Test
+    @Throws(Exception::class)
+    fun testStrip() {
+        val reference = "my string"
+        val testStrings = arrayOf(
+            reference,
+            " $reference ",
+            " $reference  ",
+            " \t $reference \t ",
+            " \n  $reference \n ",
+            "\tmy \n string\t"
+        )
+        testStrings.map { formatter.strip(it) equals reference }
+    }
+```
+
+A glimpse at the snippet above shows that while it is Kotlin code, there are some
+familiar elements from regular Java already, like package name and import statements.
+Further decomposing it reveals lack of semicolons in the end of the lines and differences
+in variable and function declarations. But also some obvious similarities like code block
+separation by usage of curly brackets and support for annotations.
```

Comment:
"But **there are** also some obvious similarities, like code block separation ~~by usage of~~ **with** curly brackets, ~~and~~ as well as ...". (Using "and" instead of ", as well as" may make it read that code blocks can be separated by curly brackets **and** by support for annotations.)